text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,908 @@
+the main exception here are future types which need to run to completion, like
+wrappers around IO completion operations. These are however unsafe by nature
+and require careful implementation and review. Therefore the `unsafe` attribute
+added here is justified.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The proposal presented here provides a way to add run to completion semantics
+for asynchronous functions in a backwards compatible and minimally invasive way.
+Since neither the way an async function desugars into a generator/state-machine nor
+the way wakeups (the task system) are performed are touched, the effort to integrate
+this mechanism should be manageable.
+
+The mechanism will allow us to provide true zero cost abstractions for completion
+based operations. Those already exist in other languages (C++, Zig, C#, Kotlin,
+etc), but not yet in Rust.
+
+`#[completion] async fn` will also introduce an async function type which behaves
+more similar to normal functions than current `async fn`s do. It might therefore
+have been interesting to see `#[completion] async fn` as the new "default" async
+function type - which might be the mostly used function type used by application
+writers in the end. However this is not possible anymore, since `async fn` is
+already stabilized.
+
+And while the introduction of another type of async function type sounds like
+as an additional complexity at first, it also has its benefits:
+
+Users will continue to be able to write simple and safe `Future` implementations
+with cancel-anytime semantics manually, and be able to use them in combination
+with powerful flow-control macros like `select!` and `join!`. A lot of types
+do not require run-to-completion semantics - e.g. async Channels, Mutexes and
+Semaphores and Timers do very well without them. We might even prefer to cancel
+those operations at any time. Only the operations which require run-to-completion
+semantics can depend on them - and are still able to use other operations
+internally.
+
+Therefore the proposal could also be seens as "providing the best of both worlds":
+Users are still able to make use of synchronous cancellation in places where it
+is safe and makes sense - which is a feature unique to Rusts `async fn`
+implementation. However they will also have to be ability to specify that their
+code needs to be run to completion if this is required for correctness of their
+application or library.
+
+## Omission of cancellation in the new `RunToCompletionFuture` type
+
+The proposal outlined here does not prescribe how cancellation is performed.
+It only mandates a "cancellation signal", and recommends cancelled asynchronous
+functions to observe and react to those.
+
+Alternative proposals, like the one in
+https://internals.rust-lang.org/t/pre-pre-rfc-unsafe-futures, also proposed to
+directly add a `.cancel()` method on the new `Future` type which needs to initiate
```

Comment:
Not saying this is a bad idea at all - but cancel and drop are completely different in my mind. Since drop is on the panic path, it simply attempts a best effort cleanup (dropping file descriptors), and protects against memory unsafety.

Cancel OTOH is highly cooperative - and on the expected path. Say you have a search problem where three tasks race to find a solution. Once one is found, the other two are cancelled - gracefully.

Reply:
Good point.  I suspect the implementations will often coincide, but that is by no means guaranteed.